프로젝트 관리 앱 [STEP2-2] 애플사이다

just1103 opened the pull request:
흰 안녕하세요. @daheenallwhite 리뷰이 애플사이다입니다. :smile: 
추가 기능을 구현하면서 RxSwift를 적용해봤습니다.

하다보니 의존성이 꼬이고, ViewModel을 제대로 활용하지 못하고 있는 것 같아요. 
STEP2-2도 많은 조언 부탁드려요.

## 구현 내용
이번 STEP2-2에서는 RxSwift를 적용하면서 Cell을 탭했을 때 Edit 화면을 띄우기, Long Press 제스처로 Popover 띄우기, Cell Swipe로 삭제 등의 추가 기능을 구현했습니다.
![Simulator Screen Recording - iPad Pro (11-inch) (3rd generation) - 2022-03-16 at 20 10 28](https://user-images.githubusercontent.com/70856586/158585828-71fb863d-05c1-4c68-b5d0-8a9f6ae887b3.gif)

### 1. 코드 구조
- 구현한 화면은 크게 Task 목록 화면 (TaskListView), Task 상세 화면 (TaskDetailView) 2개입니다.
- `TaskListViewController`에 3개 TableView가 들어있다보니 코드량이 늘어나서, 최대한 ViewModel 및 다른 코코아 클래스 (TaskTableViewCell, TaskTableView, TaskTableHeaderView)에 기능을 분리시키려고 노력했습니다.
- ViewModel은 `TaskListViewModel` 1개만 생성한 상태입니다. 저번 PR에서 View 1개당 ViewModel 1개를 만드는 게 좋다고 조언주셨는데.. 이번 프로젝트 특성상 TaskListView에 모든 정보가 다 들어있다보니 ViewModel 분리가 어렵게 느껴지는 것 같아요. (아래 궁금한 점에 고민을 정리해두었습니다.)

### 2. RxSwift 적용
- 기존에는 [Task] 타입의 배열이었던 ViewModel의 todoTasks/doingTasks/doneTasks을 `BehaviorSubject` 타입으로 변경했습니다. (이번 프로젝트에서는 Subject가 상태를 가져야 하므로 PublishSubject를 사용할 수 없다고 판단했습니다.)
- ViewModel의 Create 메서드를 호출하면 새로운 Task를 `taskRepository`의 entireTasks에 append하고, 다시 `onNext`를 통해 entireTasks (append가 완료된 상태)를 받아와서 TodoTasks (BehaviorSubject)에 전달하도록 했습니다.
- 종류별 Task의 count (ex. todoTasksCount)는 todoTasksObservable에서 stream을 따와서 `Observable`로 만들었습니다. TaskListViewController에서 HeaderView를 초기화할 때 이 Observable을 생성자 주입하도록 했습니다.
- `tableView.rx.item`을 활용하여 기존 TableView DataSource를 대체했습니다.
- RxDataSource의 `tableView.rx.didSelectItem`, `TableView.rx.tableHeaderView`도 써보고 싶었는데, 너무 어려워서 진행하지 못했습니다. :cry: 

## 고려한 점 및 궁금한 점
### 1. HeaderView 구현 및 ViewModel 분리
* Rx를 구현하기는 어려웠는데, 적용해보니 확실히 binding 덕분에 Task 변동사항을 일일이 추적하지 않아도 되어 좋았습니다.
* 그런데 현재 생성한 View가 크게 Task 목록 화면 (TaskListView), Task 상세 화면 (TaskDetailView) 2개로 나뉘는데.. TaskListView의 경우 담고 있는 내용이 워낙 많아서, ViewController는 여러 개로 분리했지만 ViewModel을 분리하기가 애매했습니다. 
* 예를 들어 `HeaderView`를 위한 ViewModel을 새로 생성한다면, 기존의 TaskListViewModel과 모르도록 해야할 것 같은데 (TaskRepository만 가지도록), 사실 `HeaderView에 필요한 ViewModel`은 이미 `TaskListViewModel`에 모두 구현되어있어서 분리하기가 애매했습니다.
* 그래서 현재는 HeaderView를 초기화할 때 `생성자 주입`으로 기존 TaskListViewModel의 일부 데이터 (taskListViewModel.todoTasksCount)를 받아오도록 했는데... 제가 생각해도 의존성이 높아져서 바람직하지 않은 것 같아요.
* 이외에도 HeaderView 자체를 현재 `TaskListViewController`가 알고 있는데, `TaskTableView` 내부에서 초기화하면 중복코드 (3개 TableView를 위해 각각의 HeaderView를 생성&등록하는 과정)를 방지할 수 있을 것 같아요. 그런데 이렇게 하면 TaskTableView 타입도 ViewModel을 알아야 하니까.. 둘 중 더 나은 방법이 뭔지 판단하기 어려웠습니다.

### 2. Long Press 제스처로 Popover 띄우기
* 위 내용과 동일한 맥락인데, 만약 `TaskListViewController`에서 LongPressGestureRecognizer를 추가하면, Recognizer를 3개 만들어서 3개 Table에 일일이 add 해줘야 하므로 비효율적이라 판단했습니다. 
* 그래서 `Cell 클래스`에 GestureRecognizer를 만들어줬는데, 막상 popover를 띄우는 건 `ViewController`의 역할이라 난감했습니다. 결국 ViewController에서 cell의 setup 메서드를 호출할 때 self(ViewController)를 넘겨주었는데.. 이것도 바람직 하지 않은 것 같아 ViewModel에서 처리해주고 싶었는데, 방법을 찾지 못했어요. 흰의 의견이 궁금합니다.
* (그리고 popover를 통해 processStatus를 변경하면 다시 Repository에 반영하기 위해 또 `ViewModel`도 필요하게 되어 ViewController로부터 전달받았는데, 이것도 ViewModel로 로직을 옮기면 해결될 수 있을 것 같아요.)

### 3. Cell 탭하면 Edit 화면을 띄우기
* `+버튼` (Task 추가 화면)과 기능이 매우 유사합니다. 원래는 공통기능을 가지는 SuperClass를 만들고, 두 화면은 각각 SubClass를 만들어 구현하려고 했는데, leftBarButton 외에는 모두 동일하기 때문에 기존의 `TaskDetailController`에다가 `taskManagerAction` 열거형 타입의 프로퍼티를 추가하여, 현재 화면의 기능 (Add인지, Edit인지)을 구분하도록 했습니다. 이 방법이 적절한지 궁금합니다.
* taskManagerAction 열거형 내부에 `leftBarButton`을 정하는 프로퍼티를 만들었는데, 이것도 엄밀히는 ViewModel을 별도 생성해서 구현해줘야 하는걸까요?

## 구현하지 못한 부분
- 마감기한이 지난 Label은 빨간색으로 표시하는 부분은 다음 STEP에서 진행해보겠습니다.
- 오류 처리는 `Print(error)` 방식으로만 확인하고 있고, 추후 리팩토링할 예정입니다. 그리고 강제추출 옵셔널을 사용하는 부분이 많은데 이 부분도 나중에 개선해보겠습니다.

이번에 구조가 꼬이면서 기능 추가에 시간을 많이 못 쓴 것 같아요.
ViewModel 분리, 의존성 주입 (어떤 타입이 어떤 타입을 가질 것인지) 관련해서 많은 조언 부탁드려요. 감사합니다~! 🙏

just1103 committed:
refactor: main 스토리보드 파일명 변경 #4

- main -> TaskListViewController
- realm package 삭제

just1103 committed:
chore: realm 패키지 삭제 #4

just1103 committed:
refactor: update 메서드 개선 및 테스트코드 수정 #4

just1103 committed:
feat: RxSwift 패키지 추가 #6

just1103 committed:
refactor: TaskListViewModel의 종류별 Tasks에 BehaviorSubject 적용 #6

just1103 committed:
refactor: TaskListViewModel의 CRUD 메서드에 Rx 적용 #6

just1103 committed:
refactor: TaskListViewModel의 기타 메서드에 Rx 적용 #6

just1103 committed:
refactor: TaskListViewController의 TableView에 Rx 적용 #6

just1103 committed:
feat: TableView Header 추가 #6

just1103 committed:
refactor: HeaderView 의존성 주입 위치 수정 #6

just1103 committed:
refactor: HeaderView Inset 설정 #6

just1103 committed:
refactor: HeaderView 의존성 주입 위치 수정 #6

just1103 committed:
feat: Add/Edit 모드에 대응하도록 TaskDetailController 기능 추가 #6

just1103 committed:
feat: TaskListViewModel에 didSelectRow 메서드 추가 #6

just1103 committed:
feat: applyTaskToEditIfExists 메서드 추가 #6

just1103 committed:
fix: applyTaskToEditIfExists 메서드 관련 초기화 이슈 해결 #6

just1103 committed:
feat: AlertFactory 타입 추가 #6

just1103 committed:
feat: LongPress 제스처에 대해 Popover 나타내는 기능 추가 #6

just1103 committed:
feat: Popover를 통해 ProcessStatus를 변경하는 기능 추가 #6

just1103 committed:
chore: 주석 삭제 #6

just1103 committed:
chore: 주석 삭제 #6

just1103 committed:
chore: 주석 추가 #6

just1103 committed:
feat: 접근제어자 추가 #6

just1103 committed:
feat: 접근제어자 추가 #6

just1103 committed:
feat: 접근제어자 수정 #6

just1103 committed:
feat: DueDate가 빠른 순으로 Task를 정렬하는 기능 추가 #6

just1103 committed:
refactor: PopoverPresenterDelegate 프로토콜 추가 #6

just1103 committed:
fix: edit 메서드의 doneTask 연결 누락부 수정 #6

just1103 committed:
feat: TaskManagerAction 타입에 rightBarButton

just1103 committed:
chore: 주석 제거 #6

just1103 committed:
feat: Cell을 Swipe하면 삭제하는 기능 추가 #6

daheenallwhite commented on ProjectManager/ProjectManager/View/TaskListView/TaskListViewController.swift:
```diff
@@ -19,97 +22,113 @@ final class TaskListViewController: UIViewController {
-        
         tableViews.forEach { tableView in
-            tableView?.dataSource = self
+            tableView?.setupTableViewCell()
```
세 개의 테이블 뷰를 담은 배열은 세팅할 때 이외에는 사용되는 곳이 없는데 세팅만을 위해 굳이 필요한 코드인가요?
storyboard 에서도 delegate 를 설정할 수 있고, processStatus 를 설정하는 과정이랑 합칠 수 있지 않나요?

just1103 replied:
일단 중복을 최소화하기 위해 배열을 만들기는 했는데, 말씀하신 것처럼 사용 빈도가 낮아서 삭제했습니다.
그리고 TableView의 setup 메서드를 통해 일괄처리하도록 변경했습니다.

just1103 replied:
일단 중복코드를 줄이려고 배열을 만들었는데, 말씀하신 것처럼 사용 빈도가 낮아서 삭제했습니다.
그리고 TableView의 setup 메서드에서 일괄처리하도록 개선했습니다.

daheenallwhite commented on ProjectManager/ProjectManager/View/TaskListView/TaskTableHeaderView.swift:
```diff
@@ -0,0 +1,82 @@
+import RxCocoa
+
+class TaskTableHeaderView: UITableViewHeaderFooterView {
+    var taskCountObservable: Observable<Int>
```
헤더뷰에서 taskCountObservable 를 가지고 있어야 할 필요가 있나요?
처음 세팅 과정에서 binding 만 하면 되지 않나요? 가지고 있어야 할 상태값인지 다시 생각해 보면 좋을 것 같아요

just1103 replied:
TaskListViewModel을 가지고 있는 TaskListViewController의 코드량이 많아져서 코드를 최대한 분리하고, 드가 3번 반복되는 것을 피하기 위해 TaskTableHeaderView로 옮겼는데, 
생각해보니 binding은 모두 TaskListViewController에서 처리하는게 관리가 용이할 것 같네요!

daheenallwhite commented on ProjectManager/ProjectManager/View/TaskListView/TaskTableHeaderView.swift:
```diff
@@ -0,0 +1,82 @@
+        label.adjustsFontSizeToFitWidth = true
+        label.textAlignment = .center
+        label.font = .preferredFont(forTextStyle: .title1)
+        
```
view생성할 때 이런 줄바꿈은 어떤 의미인가요?

just1103 replied:
switch문의 case문 내부를 제외하고, 연산 프로퍼티나 메서드 내부에서 return 윗줄은 항상 줄바꿈을 했습니다.
반환값이 중요하니까 가독성을 고려해서 그렇게 했던 것 같아요.

daheenallwhite replied:
반환값이 이미 어떤 타입인지, 어떤 세팅을 해줬는지는 위에 타입 정의에서 알 수 있지 않나요? return 자체는 객체를 반환하는 operation 인데 이게 다른 코드보다 왜 더 중요하다는건지, 가독성에 있어서 어떤 의미를 갖는지는 잘 이해가 가지 않네요..
단순히 코드량이 많아지는 것뿐만 아니라 code flow 에 있어서 생성해서, 세팅하고 넘겨주는 동작에서 return 만 더 특별한 Operation 을 하는가에 대해 생각해보면... 줄바꿈이 필요한가요?

daheenallwhite commented on ProjectManager/ProjectManager/View/TaskListView/TaskTableHeaderView.swift:
```diff
@@ -0,0 +1,82 @@
+    }
+    
+    private func setupFrame() {
+        let height = 50.0
```
height 같은 상수 literal 은 어떻게 관리하는게 좋을까요?

just1103 replied:
Design 타입 (private enum)을 추가하여 namespace를 사용하도록 했습니다!
```swift
private enum Design {
    static let frameHeight: CGFloat = 50
}
```

daheenallwhite replied:
해당 enum 을 정의한 위치가 적절한가요? TaskListViewController파일, headerview 파일에 Design 이란 같은 이름으로 정의되어 있는데 잘못 사용될 가능성은 없나요?

just1103 replied:
- 모든 파일에서 Design으로 네이밍되어 있어서 향후 실수로 다른 파일과 혼동될 수 있을 것 같아요. enum을 사용하는 타입을 덧붙여서 `DesignForTaskTableHeaderView` 등의 네이밍으로 설정하는게 안전할 것 같아요. 그리고 관련 enum들을 Design 파일로 일괄 관리하는게 바람직할 것 같습니다.
- 현재 상태로는 private을 fileprivate으로 바꿔도 동일하게 작동하는 것 같은데, 좀더 알아보겠습니다. 🧐

daheenallwhite commented on ProjectManager/ProjectManager/View/TaskListView/TaskListViewController.swift:
```diff
@@ -19,97 +22,113 @@ final class TaskListViewController: UIViewController {
+            .asDriver(onErrorJustReturn: [])
+            .drive(todoTableView.rx.items(cellIdentifier: TaskTableViewCell.reuseIdentifier,
+                                          cellType: TaskTableViewCell.self)) { [weak self] _, task, cell in
+                cell.setup(with: task, popoverPresenterDelegate: self!, viewModel: self!.taskListViewModel)
```
1. 여기서 viewModel 을 왜 개별 cell 에게 넘겨주나요? cell 내에서는 viewmodel 을 사용하지 않던데 추후 사용 예정인가요?
개별 cell 은 task next event 에서 받은 task 객체만 받아서 update 해주면 되지 않나요?
2. handler method 네이밍 관련 - 해당 함수는 next event 가 받으면 호출되는거 아닌가요? set up 이면 한번만 불리는 세팅 method 의미같은데 의도한게 맞나요?

just1103 replied:
1. cell의 presentPopover 메서드에서 Task의 processStatus를 업데이트할 때, `viewModel`의 edit 메서드를 호출해야 해서 부득이 넘겨줬습니다. 고민해봤는데.. 만약 popoverPresenterDelegate를 계속 사용한다면, delegate에게 시키는 방법도 가능할 것 같은데 적절한 방법일까요? 🤔
2. setup 메서드에는 업데이트된 Task 정보를 cell label에 반영해주는 로직이 포함되어 있습니다. 말씀하신 것처럼 그냥 고정값을 세팅하는 의미로 오인될 수 있겠네요..! `update` 네이밍으로 개선해봤습니다.

daheenallwhite replied:
### 1번 질문 관련
- update method 의 argument 타입: delegate 를 conform 한 객체가 viewcontroller 이긴 하지만 타입은 delegate로 명시하는게 더 적절하지 않을까요?
- delegate 객체를 넘기는 것 보다 viewcontroller 에서 alert 를 띄우는 함수를 핸들러로 넘겨주는건 어떨까요?

just1103 replied:
조언주신 것처럼 MVVM-C 패턴을 적용해봤고, Coordinator가 alert를 띄우도록 개선했습니다.
(Cell은 ViewModel을 알고 있고, ViewModel에 구현된 actions 클로저를 통해 처리했습니다.)

daheenallwhite commented on ProjectManager/ProjectManager/View/TaskListView/TaskListViewController.swift:
```diff
@@ -19,97 +22,113 @@ final class TaskListViewController: UIViewController {
+                  return
+              }
+
+        let taskDetailController = taskListViewModel.didSelectRow(at: indexPath.row, inTableViewOf: selectedProcessStatus)
```
didSelectRow 라는 method 네이밍이 적절한가요? 저는 method 이름만 봤을 때는 단순히 delegate 느낌의 event 를 전달하는 줄 알았는데 view controller 까지 return 하네요

just1103 replied:
위에서 남겨주신 것처럼 ViewModel에서 ViewController를 반환하는 게 부적절한것 같아서.. 더 고민해보겠습니다.

daheenallwhite commented on ProjectManager/ProjectManager/View/TaskListView/TaskListViewController.swift:
```diff
@@ -19,97 +22,113 @@ final class TaskListViewController: UIViewController {
-            cell.applyDate(with: taskListViewModel.doneTasksObservable?.value[indexPath.row])
-        default:
-            print(TableViewError.invalidTableView.description)
+        let deleteAction = UIContextualAction(style: .destructive, title: "Delete") { [weak self] _, _, handler in
```
action 생성하는 부분은 개별 method 로 분리하면 어떨까요?

just1103 replied:
createSwipeDeleteAction 메서드로 분리했습니다. 가독성이 개선된 것 같아요.

daheenallwhite reviewed (commented):
Rx 적용하느라 수고 많으셨어요 👏👏👏 짧은 시간 안에 이정도 퀄리티를 내다니 정말 대단하네요
subject, observable 등에 대한 기본적인 개념은 이미 이해하고 계신 걸로 보이네요. 크흐..👍👍
그래서 좀더 mvvm 구현에 초점 맞춰서 리뷰 코멘트 남겼어요! 에러 핸들링이라던지 자잘자잘한 포인트도 있긴 한데 이번 스텝에서는 mvvm에 더 집중해볼게요

### 질문 답변
1. headerview 를 위한 viewmodel 을 분리할 필요는 없는 걸로 보여요. 물론 이것도 view 니까 viewmodel 이 필요한거 아닌가 하고 생각할 수 있는데 그러기엔 너무 작은 부속(?), 구성하는 작은 뷰라서 viewmodel 이 필요한 레벨은 아닌걸로 생각합니다. 의존성이 높아져서 바람직하지 않다는 의견은 이해가 잘 되지 않네요. viewmodel 은 detailview controller 가 필요한 것 같은데 이 viewmodel 분리에 대해 고민해 보셨으면 좋겠어요.
2. setup 에서 view controller 넘겨주는 방식이 제가 생각하기엔 상호 참조를 만드신거 아닌가요? 물론 세개의 gesture recognizer 를 만드는건 비효율적일 수는 있는데 이미 view controller 에서 각 테이블뷰에 맞는 세팅을 세번씩 호출하는 과정이 있던데 거기에 같이 덧붙이면 어떨까요?상호참조하면 이거 memory leak 도 생길거에요..
3. `taskManagerAction` 을 만들어서 기능과 leftButton을 구분한 점은 좋네요. 다만 코멘트에서도 남겼는데 viewmodel이 viewcontroller 까지 생성해주는 역할을 가져야 하는지는 의문입니다.. viewmodel 을 TaskList, TaskDetail 로 분리하면 어떨까요?

daheenallwhite commented on ProjectManager/ProjectManager/ViewModel/TaskListViewModel.swift:
```diff
@@ -1,199 +1,152 @@
-    var todoTasksObservable: MockObservable<[Task]>? { get }
-    var doingTasksObservable: MockObservable<[Task]>? { get }
-    var doneTasksObservable: MockObservable<[Task]>? { get }
+    var todoTasksObservable: BehaviorSubject<[Task]>? { get }
+    var doingTasksObservable: BehaviorSubject<[Task]>? { get }
+    var doneTasksObservable: BehaviorSubject<[Task]>? { get }
+    var tasksObservables: [BehaviorSubject<[Task]>?] { get }
```
+ subject 에 `asObservable` 이라는 operator 가 있는데 요거 사용하면 밖에서는 event 구독만 가능할 수 있어요

just1103 replied:
asObservable을 찾아보니 `ObservableType`의 익스텐션으로 구현되어 있고, Observable을 반환해주는 기능인 것을 확인했습니다. 보통 Subject/Relay를 받아서 Observable에 할당하는 상황에 사용한다고 하는데, 아래 같은 형태를 말씀하신거죠?
```swift
let todoTasks = BehaviorSubject<[Task]>(value: taskRepository.todoTasks)
lazy var todoTasksCount: Observable<Int> = todoTasks.asObservable().map { $0.count } 
```
* 그런데 asObservable을 쓰지 않더라도 todoTasksCount을 타입 추론을 해보면 Observable인데, asObservable을 쓰면 어떤 장점이 있는지 궁금합니다. 보다 안전하게 처리를 해줄 수 있는걸까요❓

daheenallwhite replied:
네 바인딩하는 곳에서 onNext 등을 사용해서 event 를 주입할 수 있는 가능성을 완전히 배제할 수 있어요. 온전히 이벤트만 받아보고 싶을 때 요렇게 변환해서 사용하는게 좋아요

just1103 commented on ProjectManager/ProjectManager/Utilities/AlertFactory.swift:
```diff
@@ -0,0 +1,16 @@
+import UIKit
+
+struct AlertFactory {
+    func createAlert(style: UIAlertController.Style = .actionSheet,
```
Factory 패턴에서는 주로 static 메서드를 사용한다는 점을 간과했던 것 같아요.
static 메서드로 변경하면서 구조체를 열거형으로 변경했습니다!

just1103 committed:
refactor: Factory 패턴에서 Enum의 static 메서드를 사용하도록 개선 #6

just1103 committed:
refactor: 불필요한 옵셔널 타입 삭제 #7

just1103 commented on ProjectManager/ProjectManager/ViewModel/TaskListViewModel.swift:
```diff
@@ -1,199 +1,152 @@
     
     // MARK: - Methods
     func create(task: Task) {
+        taskRepository?.create(task: task)
```
이 부분도 non-optional로 수정했습니다. 👍

just1103 committed:
refactor: 프로퍼티 네이밍의 Observable 일괄 삭제 #7

just1103 committed:
refactor: TaskTableView의 setup 메서드 개선을 통해 중복코드 최소화 #7

just1103 committed:
refactor: HeaderView의 taskCountLabel binding 처리를 TaskListViewController에서 하도록 변경 #7

- TaskListViewController에 HeaderView 프로퍼티 추가, setupHeaderViewsBinding 메서드 추가
- TaskTableHeaderView의 taskCount 프로퍼티 삭제

just1103 committed:
refactor: cell의 메서드 네이밍 변경 (setup -> update) #7

just1103 commented:
흰, 기초적인 부분인데 놓친 것부터 전체적인 코드 구조까지 꼼꼼히 피드백주셔서 정말 감사해요. 😄 
ViewModel이나 타입 분리에 대해 확실히 잘 이해하게 된 것 같아요..!! 정말 많이 배우고 있습니다. ㅎㅎ

2번 답변에 대해 추가 질문이 있습니다.
> setup 에서 view controller 넘겨주는 방식이 제가 생각하기엔 상호 참조를 만드신거 아닌가요? 물론 세개의 gesture recognizer 를 만드는건 비효율적일 수는 있는데 이미 view controller 에서 각 테이블뷰에 맞는 세팅을 세번씩 호출하는 과정이 있던데 거기에 같이 덧붙이면 어떨까요? 상호참조하면 이거 memory leak 도 생길거에요..

* delegate 패턴으로 cell에서 weak var로 `popoverPresenterDelegate` 프로퍼티를 생성하여 viewController를 가지도록 개선해봤습니다. 이렇게 하면 Strong Reference Cycle이 방지된다고 알고있는데, 그럼에도 상호참조이기 때문에 지양하는 것이 좋을까요?
* ViewController에 구현하면 gesture recognizer 중복코드 외에도 popover를 띄울 때 cell이 알고있는 currentProcessStatus을 통해 변경할 ProcessStatus를 간단히 지정하는 로직 (cell의 presentPopoverToChangeProcessStatus 메서드)을 못쓰니까 코드가 길어질 것 같아서 판단이 어렵네요.. 이런 경우 어떤 기준으로 결정하는게 좋을까요?

just1103 committed:
refactor: TaskListViewModel의 메서드 네이밍 변경 (createViewControllerWithSelectedRow) #7

just1103 committed:
refactor: TaskManagerAction 열거형의 연산 프로퍼티를 DetailViewModel로 이동 #7

just1103 committed:
chore: comma 누락 수정 #7

just1103 committed:
fix: TaskDetailViewController init 수정 #7

just1103 committed:
refactor: asObservable operator 추가 #7

just1103 committed:
refactor: taskListViewModel의 createViewControllerForSelectedRow 메서드 추가 #7

just1103 committed:
refactor: createSwipeDeleteAction 메서드 추가 #7

just1103 committed:
refactor: TaskListViewController namespace 추가 #7

just1103 committed:
processStatus 열거형의 processStatusChangeOptions 프로퍼티를 ViewModel로 이동 #7

daheenallwhite commented:
빠르게 피드백 반영하셨네요~ 수고하셨습니다😉
질문에 대한 답변은 해당 코드 부분 코멘트에 남겼어요 확인부탁드려요

just1103 committed:
refactor: ViewModelProtocol을 Input/Output으로 구분 #7

just1103 committed:
refactor: MVVM-C 패턴 적용하여 FlowCoordinator 타입 추가 #7

- SceneDelegate에서 FlowCoordinator 초기화
- FlowCoordinator 타입에서 전체 ViewController 간 화면전환 구현
- TaskListViewModelActions 타입 추가하여 화면전환 작업 전달
- 각 ViewModel이 TaskListViewModelActions를 갖도록 설정
- 각 ViewController는 ViewModel의 Actions를 통해 화면전환하도록 요청

just1103 committed:
refactor: TaskListViewModelActions에 presentPopover 메서드 추가 #7

just1103 committed:
feat: TaskListViewModel에 changeDateLabelColorIfExpired 메서드 추가 #6

just1103 committed:
style: 줄바꿈 컨벤션 변경 #7

just1103 committed:
refactor: taskRepository 초기화 시점 수정 #7

just1103 committed:
feat: UseCase 추가 (TableView binding 문제 발생) #7

just1103 committed:
refactor: UseCase 삭제 #7

just1103 committed:
refactor: TaskRepository의 entireTasks 타입 변경 ([Task]->BehaviorRelay) #7

just1103 committed:
refactor: TaskListViewModel의 종류별 Tasks가 Repository의 Tasks를 subscribe 하도록 변경 #7

just1103 committed:
refactor: Repository/ViewModel의 Task 타입 변경, rx.modelSelected/rx.modelDeleted 적용 #7

- TaskRepository의 entireTasks를 BehaviorRelay 타입으로 변경
- TaskListViewModel/TaskDetailViewModel의 종류별 Task를 Observable 타입으로 변경
- TaskListViewController에서 rx.modelSelected/rx.modelDeleted 적용하여 TableView Delegate 메서드인 didSelectTask/trailingSwipeActionsConfigurationForRowAt 대체

just1103 committed:
chore: 주석 삭제 #7

just1103 committed:
chore: 주석 수정 #7

just1103 committed:
refactor: TypeOfViewController의 case 네이밍 변경 #7

daheenallwhite commented on ProjectManager/ProjectManager/Presentation/TaskListView/ViewModel/TaskListViewModel.swift:
```diff
@@ -0,0 +1,106 @@
+    let taskRepository: TaskRepositoryProtocol
+    
+    // Repository의 BehaviorRelay를 map으로 받아서 Observable로 만들어줌
+    lazy var todoTasks: Observable<[Task]> = taskRepository.entireTasks.map { tasks in
```
1. lazy 가 필요한 이유는 뭔가요?
2. TaskRepository 안에서 각 task 별로 필터링하는게 낫지 않나요?

just1103 replied:
1. taskRepository를 이니셜라이저에서 받아오기 때문에 lazy가 필요했습니다.
2. TaskRepository를 가장 단순한 형태로 가져가는 게 낫다고 판단했었는데, 말씀하신 것처럼 2개 ViewModel에서 어차피 종류별 Task를 필요로 하니까 개선해봤습니다. ViewModel에서는 Repository의 종류별 Task를 그대로 받아오도록 했습니다.
```swift
lazy var todoTasks: Observable<[Task]> = taskRepository.todoTasks
```

daheenallwhite commented on ProjectManager/ProjectManager/Presentation/TaskListView/ViewModel/TaskListViewModel.swift:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    func changeDateLabelColorIfExpired(with date: Date) -> UIColor {
+        let dayInSeconds: Double = 3600 * 24
```
상수 관리 따로 해주면 어떨까요?

just1103 replied:
별도 namespace 타입을 추가해서 개선해보겠습니다.

daheenallwhite commented on ProjectManager/ProjectManager/Presentation/TaskListView/View/TaskListViewController.swift:
```diff
@@ -0,0 +1,139 @@
+            .asDriver(onErrorJustReturn: [])
+            .drive(todoTableView.rx.items(cellIdentifier: TaskTableViewCell.reuseIdentifier,
+                                          cellType: TaskTableViewCell.self)) { [weak self] _, task, cell in
+                cell.update(with: task, viewModel: self!.taskListViewModel)
```
optional unwrapping 이 필요한가요? view model 이 optional 타입일 필요가 있나요?

just1103 replied:
해당 ViewController를 초기화할 때, ViewModel을 생성자 주입으로 전달하고 싶었고, Storyboard를 사용하다보니 `init?(coder: NSCoder)`을 써야 해서 불가피하게 ViewModel 타입을 옵셔널로 지정했습니다. 
다른 방법이 있을까요?

daheenallwhite commented on ProjectManager/ProjectManager/Presentation/TaskListView/View/TaskListViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        
+        // 기존 TableView Delegate의 didSelectTask 메서드를 대체
+        todoTableView.rx.modelSelected(Task.self)
+            .bind(onNext: { self.taskListViewModel.actions?.showTaskDetailToEditTask($0) })
```
위에 handler 에는 weak self 가 있고 여기는 없는데 이유가 있나요?

just1103 replied:
누락한 부분 개선했습니다!

daheenallwhite commented on ProjectManager/ProjectManager/Presentation/TaskListView/View/TaskListViewController.swift:
```diff
@@ -0,0 +1,139 @@
+             }
+             .disposed(by: disposeBag)
+        
+        // 기존 TableView Delegate의 didSelectTask 메서드를 대체
```
이 method 가 크게는 세 부분으로 나뉘어 지는 걸로 보이는데 메소드로 분리해주면 어떨까요? comment 로 부분을 지정하는 것 보다 코드로 이야기하는게 더 적절하지 않을까요?

just1103 replied:
분리해주었습니다~!

daheenallwhite commented on ProjectManager/ProjectManager/Presentation/TaskListView/View/TaskTableViewCell.swift:
```diff
@@ -0,0 +1,76 @@
+        task = nil
+    }
+    
+    func update(with task: Task, viewModel: TaskListViewModelProtocol) {
```
TaskListViewModel 에 너무 많은 역할이 있는 것 같아요. 컬러 값을 가져온다던지 하는게... Task 하나를 표현하는 TaskViewModel 로 분리해보면 어떨까요?

just1103 replied:
TaskViewModel 타입을 생성하여 분리했습니다. TaskList에 해당하지 않는 `개별 Task`를 관리하는 기능을 배치하면 유용할 것 같아요. 👍

daheenallwhite commented on ProjectManager/ProjectManager/Presentation/TaskDetailView/View/TaskDetailController.swift:
```diff
@@ -0,0 +1,89 @@
+import UIKit
+
+class TaskDetailController: UIViewController {
+    private var taskDetailViewModel: TaskDetailViewModelProtocol?
```
요기는 그냥 optional 타입이네요. TaskListViewController 랑 viewmodel 을 다르게 관리하는 이유가 있나요?

just1103 replied:
Rx에 신경쓰다가 놓친 부분인데, 일관성 있도록 암시적 추출 옵셔널을 사용하지 않도록 수정하겠습니다!

daheenallwhite commented on ProjectManager/ProjectManager/Presentation/TaskDetailView/ViewModel/TaskDetailViewModel.swift:
```diff
@@ -0,0 +1,80 @@
+    // MARK: - Properties
+    let taskRepository: TaskRepositoryProtocol
+    
+    lazy var todoTasks: Observable<[Task]> = taskRepository.entireTasks.map { tasks in
```
이 연산은 viewmodel 마다 중복되니까 위에서 질문한대로 task repository 에서 해주는게 나을 것 같아요

just1103 replied:
개선 완료했습니다.

daheenallwhite reviewed (commented):
버그 고치느라 고생하셨네요
rx 기능 사용에 대해서는 전반적으로 다 이해하고 계신 것 같아요
다만 역할 분리랑 mvvm 에서 어색한 점이 몇군데 있어서 코멘트 새로 남겼어요~

just1103 committed:
refactor: TableView binding 관련 메서드 분리 #7

just1103 committed:
refactor: TaskListViewController 옵셔널 타입 및 약한 참조 관련 개선 #7

just1103 committed:
refactor: TaskViewModel 타입 추가 #7
